### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by Azaryann**

### CODE REVIEW

File: presentation/src/main/java/am/smartcafe/presentation/controller/RegisterController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public ModelAndView registration() {
+        ModelAndView modelAndView = new ModelAndView();
+        User user = new User();
+        modelAndView.addObject("user", user);
+        modelAndView.setViewName("register");
+        return modelAndView;
+    }
+
+    @PostMapping("/register")
+    public ResponseEntity <?> createNewUser(@RequestBody UserRegisterRequest userRegisterRequest, BindingResult bindingResult) throws ModelAlreadyExistException, ModelAlreadyExistException {
```
Comment (SevoMkrtchyan) [inline presentation/src/main/java/am/smartcafe/presentation/controller/RegisterController.java]: ModelAlreadyExistException used 2 time, please change it

Comment (Azaryann) [reply presentation/src/main/java/am/smartcafe/presentation/controller/RegisterController.java]: yes, changed...

File: data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+import am.smartcafe.data_access.model.Role;
+import lombok.Builder;
+
+import javax.validation.constraints.*;
+import java.util.Objects;
+import java.util.Set;
+
+@Builder
+public class UserRegisterResponse {
```
Comment (SevoMkrtchyan) [inline data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: do we need validation on response?

File: data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    @Size(min = 2, max = 50)
+    @NotEmpty(message = "*Please provide your last name")
+    private String lastName;
+
+    @NotBlank
+    private Set<Role> role;
+
+    @Column(name = "active")
+    private int isActive;
```
Comment (SevoMkrtchyan) [inline data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: state will be set with user service

File: data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    @Email
+    @Pattern(regexp = "^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$", message = "Please enter a correct email")
+    private String email;
+
+    @NotBlank
+    @Length(min = 6, message = "*Your password must have at least 6 characters")
+    @NotEmpty(message = "*Please provide your password")
+    private String password;
```
Comment (SevoMkrtchyan) [inline data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: password is generated automatically during saving

File: data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Set;
+
+import lombok.Builder;
+
+@Builder
+public class UserRegisterRequest {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
```
Comment (SevoMkrtchyan) [inline data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: this field would be in User entity, not here

File: data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java (hunk 3 above)
Comment (Azaryann) [reply data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: I mean, we don't need isActive field here, do we?

Comment (SevoMkrtchyan) [reply data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: yes,of course

Comment (Azaryann) [reply data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: changed...

File: data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java (hunk 2 above)
Comment (Azaryann) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: UserRegisterResponse used in the UserService

Comment (SevoMkrtchyan) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: as I understand we send this response back after successfully registration right?

Comment (Azaryann) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: yes, that's how I wrote the logic

Comment (SevoMkrtchyan) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: Ok, but I think we don't need to send back validated information if it validated already, is it?

Comment (Azaryann) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: Yes, you are right, I may have made a mistake in that part

Comment (SevoMkrtchyan) [reply data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserRegisterResponse.java]: it's ok, may happen

File: service/src/main/java/am/smartcafe/util/UserMapper.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package am.smartcafe.util;
+
+import am.smartcafe.data_access.dto.req.UserRegisterRequest;
+import am.smartcafe.data_access.dto.resp.UserResponse;
+import am.smartcafe.data_access.model.User;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (SevakMart) [inline service/src/main/java/am/smartcafe/util/UserMapper.java]: @Service is redundant

File: service/src/main/java/am/smartcafe/service/UserService.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+import am.smartcafe.data_access.dto.resp.UserResponse;
+import am.smartcafe.data_access.model.User;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+public interface UserService {
+  UserResponse saveUser(User user);
```
Comment (SevakMart) [inline service/src/main/java/am/smartcafe/service/UserService.java:9]: save is enough

File: service/src/main/java/am/smartcafe/service/UserService.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package am.smartcafe.service;
+
+import am.smartcafe.data_access.dto.resp.UserResponse;
+import am.smartcafe.data_access.model.User;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
```
Comment (SevakMart) [inline service/src/main/java/am/smartcafe/service/UserService.java]: @Service is redundant

File: service/pom.xml [hunk 9]
```diff
@@ -23,5 +23,18 @@
...
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.springframework.security/spring-security-core -->
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-core</artifactId>
+            <version>5.6.0</version>
+        </dependency>
+        <dependency>
+            <groupId>am.smartcafe</groupId>
```
Comment (SevakMart) [inline service/pom.xml:28]: the service layer should not know anything about service layer

File: data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserResponse.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import am.smartcafe.data_access.model.Role;
+import lombok.Builder;
+
+import javax.validation.constraints.*;
+import java.util.Objects;
+import java.util.Set;
+
+@Builder
+public class UserResponse {
+  private static final String REGEX = "^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$";
```
Comment (SevakMart) [inline data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserResponse.java]: VALID_EMAIL_REGEX ?

File: data-access/pom.xml [hunk 11]
```diff
@@ -10,6 +10,30 @@
...
+            <artifactId>spring-boot-starter-web</artifactId>
+            <version>2.6.1</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jpa</artifactId>
+        </dependency>
+
+        <dependency>
```
Comment (SevakMart) [inline data-access/pom.xml:27]: avoid using lombok

File: data-access/pom.xml [hunk 12]
```diff
@@ -10,6 +10,30 @@
     <modelVersion>4.0.0</modelVersion>
 
     <artifactId>data-access</artifactId>
+    <dependencies>
+
+        <!-- https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-web -->
+        <dependency>
```
Comment (SevakMart) [inline data-access/pom.xml:16]: if this dependency needed more than one module you can add it into parent pom

File: data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SevakMart) [inline data-access/src/main/java/am/smartcafe/data_access/dto/req/UserRegisterRequest.java]: do you really need to override equals and hashcode ?

File: data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserResponse.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
+package am.smartcafe.data_access.dto.resp;
```
Comment (SevakMart) [inline data-access/src/main/java/am/smartcafe/data_access/dto/resp/UserResponse.java]: move this to service layer

File: presentation/src/main/resources/templates/register.html [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+                        <label
+                                th:if="${#fields.hasErrors('email')}" th:errors="*{email}"
+                                class="validation-message">
+                        </label>
+                    </div>
+                </div>
+
+                <div class="form-group">
+                    <div class="col-sm-9">
+                        <button type="submit" class="btn btn-primary btn-block">Register User</button>
```
Comment (SevakMart) [inline presentation/src/main/resources/templates/register.html:69]: please use i18n, the locale resolvers are merged to master already

File: presentation/src/main/java/am/smartcafe/presentation/SmartCafeApplication.java [hunk 16]
```diff
@@ -2,12 +2,17 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
 
+@EnableJpaRepositories(basePackages = {"am.smartcafe.data_access.repository"})
+@EntityScan({"data-access.am.smartcafe.data_access.model"})
```
Comment (SevakMart) [inline presentation/src/main/java/am/smartcafe/presentation/SmartCafeApplication.java]: Do we need data access prefix here?

Comment (Azaryann) [reply presentation/src/main/java/am/smartcafe/presentation/SmartCafeApplication.java]: Dear Sevak, at first there was a problem with writing, that annotation helped, then I did not try to remove it

**Pull request merged**